akka-actor/src/main/scala/akka/pattern/Patterns.scala

```diff
@@ -485,6 +487,51 @@ object Patterns {
       scheduler: Scheduler,
       ec: ExecutionContext): CompletionStage[T] =
     scalaRetry(() => attempt.call().toScala, attempts, delay.asScala)(ec, scheduler).toJava
+
+  /**
+   * Returns an internally retrying [[scala.concurrent.Future]].
+   * The first attempt will be made immediately, and each subsequent attempt will be made after 'delay'.
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry.
+   * You can provide a delayFunction to generate the next delay duration if the initial invocation has error occur,
+   * which accepts the current attempted count(start from `0`), and returns a [[java.util.Optional]] of next delay duration.
+   * If the [[java.util.Optional]] is present, the next attempt will be made after the delay, if the [[java.util.Optional]]
+   * is empty, then will stop any further retry and return with a failed [[scala.concurrent.Future]].
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](
+      attempt: Callable[Future[T]],
+      delayFunction: akka.japi.function.Function[Int, Optional[java.time.Duration]],
```

Comment:
I think that has to be `java.lang.Integer` instead of `Int` for Java API

Reply:
Yes

Reply:
Should we just use the `java.util.function.IntFunction`? is there any special reason that we are still using the `akka.japi.function.Function`?

Reply:
I will add the others for exponential backoff?

Reply:
`IntFunction` would be fine here, since it's not expected that this function throws checked exceptions